File: 100dynamic/279.swift
class Solution {
    func numSquares(_ n: Int) -> Int {
        var dp:[Int] = Array(repeating:Int.max, count:n+1)
        dp[0] = 0
        for i in 1...n {
            for j in stride(from: Int(sqrt(Double(i))), through: 0, by: -1) {
                dp[i] = min(dp[i - j*j] + 1, dp[i])
            }
        }
        return dp[n]
    }
}

File: 100dynamic/279.rs
impl Solution {
    pub fn num_squares(n: i32) -> i32 {
        let mut dp:Vec<i32> = vec![std::i32::MAX; (n+1) as usize];
        dp[0] = 0;
         for i in 1..=n {
            let mut j = 1;
            while j * j <= i {
                dp[i as usize] = dp[i as usize].min(dp[(i - j * j) as usize] + 1);
                j += 1;
            }
        }
        dp[n as usize]

    }
}